src/dbnode/storage/index/block.go

```diff
@@ -162,32 +471,242 @@ func (b *block) WriteBatch(inserts *WriteBatch) (WriteBatchResult, error) {
 	}
 
 	partialErr, ok := err.(*m3ninxindex.BatchPartialError)
-	if !ok { // should never happen
-		err := b.unknownWriteBatchInvariantError(err)
-		// NB: marking all the inserts as failure, cause we don't know which ones failed
+	if !ok {
+		// NB: marking all the inserts as failure, cause we don't know which ones failed.
 		inserts.MarkUnmarkedEntriesError(err)
 		return WriteBatchResult{NumError: int64(inserts.Len())}, err
 	}
 
 	numErr := len(partialErr.Errs())
 	for _, err := range partialErr.Errs() {
-		// Avoid marking these as success
+		// Avoid marking these as success.
 		inserts.MarkUnmarkedEntryError(err.Err, err.Idx)
 	}
 
-	// mark all non-error inserts success, so we don't repeatedly index them
+	// Mark all non-error inserts success, so we don't repeatedly index them.
 	inserts.MarkUnmarkedEntriesSuccess()
 	return WriteBatchResult{
 		NumSuccess: int64(inserts.Len() - numErr),
 		NumError:   int64(numErr),
 	}, partialErr
 }
 
-func (b *block) executorWithRLock() (search.Executor, error) {
-	var expectedReaders int
-	if b.activeSegment != nil {
-		expectedReaders++
+func (b *block) foregroundCompactWithBuilder(builder segment.DocumentsBuilder) error {
+	// We inserted some documents, need to compact immediately into a
+	// foreground segment.
+	b.Lock()
+	foregroundSegments := b.foregroundSegments
+	b.Unlock()
+
+	segs := make([]compaction.Segment, 0, len(foregroundSegments)+1)
+	segs = append(segs, compaction.Segment{
+		Age:  0,
+		Size: int64(len(builder.Docs())),
+		Type: segments.MutableType,
+	})
+	for _, seg := range foregroundSegments {
+		segs = append(segs, compaction.Segment{
+			Age:     seg.Age(),
+			Size:    seg.Segment().Size(),
+			Type:    segments.FSTType,
+			Segment: seg.Segment(),
+		})
 	}
+
+	plan, err := compaction.NewPlan(segs, b.opts.ForegroundCompactionPlannerOptions())
+	if err != nil {
+		return err
+	}
+
+	// Check plan.
+	var planErr error
+	switch {
```

Comment:
switch statement seems a little weird to me since you're just trying to return an error. How is it different from just doing:

```golang
if len(plan.Tasks) == 0 {
    return errForegroundComapctorNoPlan
}

if !taskHasNilSegment(plan.Tasks[0]) {
    return errForegroundCompactorBadPlanFirstTask
}
...
```

Reply:
It's not, I thought I'd just do a final sanity check and group them all together.  

I'll break it up to just the default Go-ism of if err/ret, if err/ret, etc to match what you suggest.

Reply:
kk cool I think that would be good, the fact that it diverged made me stop and spend some time on it to see if it was doing something unexpected